stratum/hal/lib/common/p4_service.cc

```diff
@@ -713,168 +712,59 @@ ::util::Status P4Service::AddOrModifyController(
     // Store Channel and tid for Teardown().
     stream_response_reader_tids_.push_back(tid);
     stream_response_channels_[node_id] = channel;
-    node_id_to_controllers_[node_id] = {};
-    it = node_id_to_controllers_.find(node_id);
+    node_id_to_controller_manager_.emplace(node_id, node_id);
```

Comment:
Should we put a `SdnControllerManager` instead of a `uint64` to the map as the value?
Since the type of `node_Id_to_controller_manager_` is `std::map<uint64, p4runtime::SdnControllerManager>`

Reply:
It actually does that, but it looks confusing. Here's what's going on:

- `SdnControllerManager` cannot be copied, only moved
- hence we do `emplace` instead of `insert`
- but we still can't construct the manager object here (e.g. `emplace(node_id, SdnControllerManager())`) as that would be a copy
- so we use [perfect forwarding](https://en.cppreference.com/w/cpp/utility/forward) semantics to construct the manager directly _inside_ the map only
- as a result, here, we're only providing the constructor args of the manager to be forwarded, hence the second `node_id`.

Hope this makes sense.

Reply:
Now I get it. Totally make sense now. It is ok to leave this code here.
Maybe we can put some comments here so any other people won't confuse like me in the future?